Improved build performance for repeated builds

benjie332 opened the pull request:
## Description ##

Some Gradle tasks were configured to prevent Gradle from using up-to-date checks (and build caching, if enabled). This meant that some tasks were being executed even if the output was not different, wasting build time.


This PR proposes changes to the build scripts to make the builds more efficient for repeated builds on the same machine. This should save a few seconds, with more improvements if build caching were to be enabled.

Some key highlights:

- tasks that share output locations (eg. tasks that downloaded files directly into build/classes) would prevent Gradle from identifying the owning tasks, which disables caching/up-to-date checks. These were split into separate locations and left to the `processResources`/`jar` tasks to collect. For some modules, the file was placed directly into the source, and the version property was injected to ensure repeatability in the build  - feedback is welcome on this change.

- tasks that do not declare inputs/outputs or have no outputs also disable the checks. These tasks were updated to include the right inputs/outputs. In the case of the `verifyJava` task, a new output file was created to let Gradle know when the task was completed successfully previously. If inputs do not change, Gradle will automatically skip these tasks, as the output would have remained the same.

- absolute paths were changed to relative paths (or normalized to only consider relative paths). If remote build caches are enabled, the build cache will be used for these tasks across any connected build machine.


The improvements were discovered through analyzing the build with [Develocity](https://gradle.com/develocity/). This [Build Scan](https://scans.gradle.com/s/w2yxmjzm2ciqm) is of a second `build` run after the improvements were implemented. 

The [timeline view](https://scans.gradle.com/s/w2yxmjzm2ciqm/timeline) shows a breakdown of where time was spent in the build.  In this case, as it was performed on a machine that had already run the `build` task, you can see [here](https://scans.gradle.com/s/w2yxmjzm2ciqm/timeline?outcome=success,failed) that no tasks were executed as they were all up to date.

ribafish commented on engines/mxnet/mxnet-engine/build.gradle:
```diff
@@ -32,7 +29,12 @@ pmdMain.source = 'src/main/java'
     dependsOn ":engines:mxnet:jnarator:jar"
+    finalizedBy verifyJava
+    //declare inputs and outputs to allow Gradle to cache the task or skip it if inputs do not change
+    inputs.file("${project.projectDir}/src/main/jna/mapping.properties")
```
Add the input name here please 👍

benjie332 replied:
Done

ribafish commented on engines/mxnet/mxnet-engine/build.gradle:
```diff
@@ -33,13 +33,11 @@ pmdMain.source = 'src/main/java'
     outputs.dir "${project.buildDir}/generated-src"
 
-    //always cacheable if inputs are the same
-    outputs.cacheIf { true }
```
Why did you remove this?

ribafish replied:
What about this? Is there a reason we don't want to make this cacheable?

benjie332 replied:
I had added it myself to force caching, but it didn't seem to make a difference when I re-ran the experiments after other fixes, so I opted out of adding it to make things simpler.

ribafish replied:
Hm, unless caching of the outputs is enabled elsewhere (I couldn't find it), removing this will make this task not cacheable. Can you verify with exp2 or 3?

benjie332 replied:
Ah, I missed it because I was looking at grouping by type, but you are completely right, it becomes non-cacheable:

![image](https://github.com/benjie332/djl/assets/3155713/1eddde66-3880-4b66-9252-ebbc81a857f9)

I will readd this. I did the same to another task, I will fix it there too.

benjie332 replied:
In fact, with the 2 suggestions you made:

`Executed non-cacheable tasks:    186 tasks, 1m 10.688s total execution time`
->
`Executed non-cacheable tasks:    150 tasks, 1m 1.105s total execution time`

ribafish commented on tools/gradle/java-formatter.gradle:
```diff
@@ -47,8 +48,7 @@ class JavaFormatterPlugin implements Plugin<Project> {
                     .withPathSensitivity(PathSensitivity.RELATIVE)
             outputs.file(project.file(resultFilePath))
-            //is always cacheable if inputs are the same
-            outputs.cacheIf { true }
```
Why did you remove this?

ribafish replied:
What about this? Is there a reason we don't want to make this cacheable?

benjie332 replied:
Same reason as above

benjie332 replied:
I've since readded it 😁

ribafish commented on engines/llama/build.gradle:
```diff
@@ -43,15 +44,15 @@ processResources {
         }
         copy {
             from jnilibDir
```
Did you verify whether changes to the files in this folder will trigger task execution?

benjie332 replied:
The files are versioned and downloaded from a remote URL given the input property versions.

My reasoning is since the files are versioned (and the processing is always the same), then if the input property versions do not change, the files will not either. (In fact I cannot see a way how to test this without modifying the task itself)

ribafish replied:
Hm, that's true, now that the version is tracked that should take care of it 👍

ribafish commented on engines/llama/build.gradle:
```diff
@@ -43,15 +44,15 @@ processResources {
         // write properties
-        def propFile = file("${project.projectDir}/build/classes/java/main/native/lib/llama.properties")
+        def propFile = file(baseResourcePath + "/native/lib/llama.properties")
         propFile.text = "version=${llamacpp_version}-${version}\n"
```
Did you verify whether changes to the versions specified here will trigger task execution?

benjie332 replied:
Fixed - added input properties

ribafish commented on engines/ml/xgboost/build.gradle:
```diff
@@ -39,7 +39,7 @@ dependencies {
+    def jnilibDir = "${project.buildDir}/resources/main/lib/linux/aarch64/"
     outputs.dir file(jnilibDir)
     doLast {
         def url = "https://publish.djl.ai/xgboost/${xgboost_version}/jnilib/linux/aarch64/libxgboost4j.so"
```
Did you verify whether changes to the version specified here will trigger task execution?

benjie332 replied:
Fixed, added input property

ribafish commented on engines/tensorrt/build.gradle:
```diff
@@ -31,14 +33,14 @@ processResources {
-        file("${project.buildDir}/classes/java/main/native/lib").mkdirs()
+        file("${baseResourcePath}/native/lib").mkdirs()
         copy {
             from jnilibDir
```
Did you verify whether changes to the files in this folder will trigger task execution?

benjie332 replied:
Same reasoning as previous - the files would not change if the properties do not

ribafish commented on engines/tensorrt/build.gradle:
```diff
@@ -31,14 +33,14 @@ processResources {
         // write properties
-        def propFile = file("${project.buildDir}/classes/java/main/native/lib/tensorrt.properties")
+        def propFile = file("${baseResourcePath}/native/lib/tensorrt.properties")
         propFile.text = "version=${trt_version}-${version}"
```
Did you verify whether changes to the versions specified here will trigger task execution?

benjie332 replied:
Fixed - added input properties

ribafish commented on api/build.gradle:
```diff
@@ -17,12 +17,8 @@ dependencies {
-        def propFile = new File(classesDir, "api.properties");
-        propFile.text = "djl_version=${project.version}"
+    filesMatching("**/api.properties") {
+        expand projectVersion: project.version
```
Did you verify whether changes to the version specified here will trigger task execution?

benjie332 replied:
Nice catch 😄 Added project.version as an input

ribafish reviewed (changes requested):
Looks good, I'd just like you to verify that changes to the files and properties used by `processResources` tasks trigger task re-execution if they change. I wrote that down in a couple of places, but didn't want to add it everywhere explicitly - all should be checked though.

runningcode commented on engines/mxnet/jnarator/build.gradle:
```diff
@@ -33,3 +33,5 @@ jar {
     from configurations.runtimeClasspath.collect { it.isDirectory() ? it : zipTree(it) }
 }
+
+generateGrammarSource.finalizedBy verifyJava
```
why is this a `finalizedBy` relationship and not a `dependsOn` from `verifyJava`?

benjie332 replied:
My reasoning here was that `verifyJava` was in a centralized gradle file, and adding all the required dependencies there would mean that over time that file could grow to know about a lot of other tasks.

As such I inverted the relationship so that the tasks that are distributed in the modules finalize the common tasks. In theory, this means that when adding a new module, you only need to worry about your module's build file, not any others. 

In theory 😄

ribafish replied:
So, if you run `verifyJava`, does the `generateGrammarSource` also run? [Here](https://discuss.gradle.org/t/when-should-one-use-dependson-and-when-finalizedby/4066/2) is a nice explanation when to use which ordering, and [here](https://docs.gradle.org/current/userguide/controlling_task_execution.html) are the docs on task ordering and depednencies.

benjie332 replied:
Ah, I see what was meant originally.... let me check to see if it should be`dependsOn`, I don't recall why I added it.

I misunderstood the comment and thought you were referring to why I added the `finalizedBy` to this task instead of where verifyJava was declared 🙃

benjie332 replied:
Fixed - there was a hidden dependency of tasks that I 'patched' with `finalizedBy`, but it really should have been a chain of `dependsOn`. 

Thanks for the comments!

runningcode commented on engines/mxnet/mxnet-engine/build.gradle:
```diff
@@ -32,7 +29,12 @@ pmdMain.source = 'src/main/java'
 
 tasks.register('jnarator') {
     dependsOn ":engines:mxnet:jnarator:jar"
+    finalizedBy verifyJava
```
why is this a `finalizedBy` relationship and not a `dependsOn` from the `verifyJava`?

benjie332 replied:
Same reasoning as the other case

benjie332 replied:
Fixed this too, it is now a proper dependency

runningcode commented on engines/onnxruntime/onnxruntime-engine/build.gradle:
```diff
@@ -24,7 +25,7 @@ processResources {
                 "token_classification",
         ]
-        def prefix = "${projectDir}/build/classes/java/main/nlp"
+        def prefix = "${baseResourcePath}/nlp"
```
you can move this definition to line 18 and then re-use it to specify `outputs.dir`

benjie332 replied:
Thanks, updated

runningcode commented on engines/pytorch/pytorch-engine/build.gradle:
```diff
@@ -14,12 +14,8 @@ dependencies {
-        classesDir.mkdirs()
-        def propFile = new File(classesDir, "pytorch-engine.properties");
-        propFile.text = "djl_version=${djl_version}\npytorch_version=${pytorch_version}"
+    filesMatching("**/pytorch-engine.properties") {
```
was the `outputs.file` intentionally removed?

benjie332 replied:
In this case yes, it is part of the source tree now instead of being generated. (It is only the properties that are expanded).

There seemed to be 2 types of process resources tasks:
- this type, which was generating a version file. I simplified these to inject properties into a committed source file instead
-  another type which downloaded file(s) from URL(s) and processed them to be included in some way. Those I left as-is, just declaring inputs/outputs accordingly.

runningcode reviewed (commented):
nice fixes! i left some comments :)

ribafish commented on engines/mxnet/mxnet-engine/build.gradle:
```diff
@@ -32,7 +30,14 @@ pmdMain.source = 'src/main/java'
+        .withPathSensitivity(PathSensitivity.RELATIVE)
+        .withPropertyName("jna/mapping.properties file")
     outputs.dir "${project.buildDir}/generated-src"
+    //there is no reason not to cache the task if Gradle needs to, so enable it
```
nit: I think you can remove this comment

runningcode replied:
i agree we can remove this. how long does this task take to execute?

ribafish commented on tools/gradle/java-formatter.gradle:
```diff
@@ -39,6 +39,18 @@ class JavaFormatterPlugin implements Plugin<Project> {
+                    .withPropertyName('generatedSourceFiles')
+                    .withPathSensitivity(PathSensitivity.RELATIVE)
+            outputs.file(project.file(resultFilePath))
+            //there is no reason not to cache the task if Gradle needs to, so enable it
```
nit: I think you can remove this comment

ribafish reviewed (approved):
Some nitpicks, otherwise it looks good to me 👍

runningcode commented on engines/mxnet/mxnet-engine/build.gradle:
```diff
@@ -55,6 +60,8 @@ tasks.register('jnarator') {
     }
 }
 
+verifyJava.dependsOn jnarator
```
can we define all `verifyJava` dependencies in the `verifyJava` tasks definition? this would consolidate all the logic related to verifyJava in one place

runningcode commented on engines/mxnet/jnarator/build.gradle:
```diff
@@ -33,3 +35,7 @@ jar {
     from configurations.runtimeClasspath.collect { it.isDirectory() ? it : zipTree(it) }
 }
+
+verifyJava.dependsOn generateGrammarSource
```
i added the same comment below but can we consolidate all the logic related to verifyJava in one place?

another point, we should use the task configuration avoidance API. this would mean it would look like `dependsOn(tasks.named(generateGrammarSource))`

See more information on task configuration avoidance here: https://docs.gradle.org/current/userguide/task_configuration_avoidance.html

Note: please do not go in and update all the uses of the eager task configuration, that would be a lot of work, we should just ensure that newly added uses are lazy.

benjie332 replied:
Hmm, I tried doing this but the way `verifyJava` is defined, it just ends up sprouting a bunch of 'not found' errors in various modules:

`Task with name 'generateTestGrammarSource' not found in project ':api'`

I assume this is because the dependencies are only relevant for a few projects which actually declare the relevant tasks, but `verifyJava` is applied globally.

runningcode reviewed (commented):
Looks great, I added some comments. Please also add a build scan link showing the improvements to the PR comment

benjie332 commented:
@runningcode @ribafish does the PR description make sense?

If there are no objections (I think I addressed all the comments), I'd like to propose the PR to the original repo

benjie332 commented:
Merged upstream! Closing this PR
